### ISSUE TITLE
Turn target name not updated ahead of some junctions

### ISSUE THREAD

**Issue by polarbearing** [osmandapp/Osmand#3007]
nb-2016-08-23

Normally the name of the street the next turn leads to is show very early, up to 1200m even on a tertiary road. On some junctions however, the current road display remains even while standing at the traffic light. The turn target is shown very briefly when already on the junction. Voice navigation is not affected, the turn and the target name are announced early enough, and pressing the turn button repeats the target name, while the street widget is still stuck. 

In this context, the blue "current road" arrow becomes confusing, it could be read as "carry on"; the crosshair icon originally discussed in #2862 would be less ambiguous. 

The examples below were experienced twice in real driving, while other junctions were correct, so the bug seems to be systematic:

![screenshot_2016-08-26-14-43-55](https://cloud.githubusercontent.com/assets/8266355/18021358/18ea8dc0-6be7-11e6-92f7-32db28517c19.png)
[Location](http://www.openstreetmap.org/?mlat=52.46059&mlon=13.50406#map=19/52.46059/13.50406), coming from Schnellerstr, turning into Karlshorster Str (observed twice)

![screenshot_2016-08-24-22-35-05](https://cloud.githubusercontent.com/assets/8266355/18021537/56d16f22-6be8-11e6-9a12-c8c43941b6d7.png)
[Location](http://www.openstreetmap.org/?mlat=52.45156&mlon=13.59285#map=19/52.45156/13.59285), coming from Friedrichshagener Str, turning into Salvador-Allende-Str (observed twice)

![screenshot_2016-08-26-14-41-06](https://cloud.githubusercontent.com/assets/8266355/18021573/a6733dc6-6be8-11e6-8feb-0ffb0eb0fcd9.png)
[Location](http://www.openstreetmap.org/?mlat=52.45307&mlon=13.51531#map=18/52.45307/13.51531), coming from Fennstr, turning into Michael-Brückner-Str (observed twice)

**Comment by CJTmmr** [comment]
Related/duplicate to/of  #2725

**Comment by vshcherb** [comment]
I see one very similar thing on all screenshots, turn lane distance is different from distance on the widget.

**Comment by sonora** [comment]
Victor's observation indicates that on these intersections there is an additional short route segment before the final turn onto the segment with the new name.

The fact that it works for voice prompts means that we properly mask the interim segments with isSpeak=false, while we keep them in the series of nav instructions for the optical lane indicator, this prevents the "new name" display to be shown until the driver is on the very last segment (which is probably the "main turn" for the driver), and which on these intersections seems to be right on the intersection.

Interesting that we get it right not only for voice prompting but also for the next turn widget. So maybe the issue can be solved by for the street name display somehow applying the logic we use for the next turn widget and not that for the lanes widget. (I think the next turn widget tries to summarize some turns, while the lanes indication considers a lot more "interim" maneuvers needed for lane selection).

**Comment by vshcherb** [comment]
Next turn widget just skips the turns with speak=false, I think here the logic should be more difficult. So we should analyze if next turn is in green area, then we should use the "next turn widget" information.

**Comment by sonora** [comment]
Then again, I am not sure how frequent and how significant this issue really is in practice.

Maybe the issue is much easier addressed by simply replacing the potentially confusing blue arrow with a crosshair icon ⌖ (like e.g. `map_my_location.png`).

**Comment by CJTmmr** [comment]
In my experience this issue is always present when:

- Intersection with divided ways   AND

- Incoming highway is tagged with turn:lanes  AND

- all ways on intersection are separate sections, (to have the right turn:lanes on each section)

Since I was rather busy tagging turn:lanes in my region, the issues is frequent in my region...

If needed I could provided multiple screenshots..

Greeting Kees-Jan

**Comment by sonora** [comment]
Yes, exactly. the "additional routing segment" I am talking about should be the one connecting the divided highway. Only once that is entered, the "new street name" is displayed.

**Comment by polarbearing** [comment]
> Maybe the issue is much easier addressed by simply replacing the potentially confusing blue arrow with a crosshair icon ⌖

independently of the turn issue, yes please.

**Comment by polarbearing** [comment]
I have a case [here](http://www.openstreetmap.org/node/26953179) that does not involve a dual carriageway. The strange thing here is that the slight-right turn is neither present in the main navigation nor in the target road. Only the turn:lanes are correct. 
![screenshot_2016-09-04-14-07-29](https://cloud.githubusercontent.com/assets/8266355/18397038/beb942de-76c6-11e6-9c01-ee35d657ff38.png)

**Comment by sonora** [comment]
Could you please test the latest nightly, I have made a small change which I think may fix the issue.

**Comment by CJTmmr** [comment]
Route Simulation does not show turnlane widget nor next-street widget.
First opportunity to test on the road will be Monday..

Op 11 sep. 2016 01:22 schreef "Hardy" <notifications@github.com>:

**Comment by sonora** [comment]
From my testing it works now. But please test all your examples, too.

I will leave the issue open for Victor (@vshcherb) to decide if we want to change the icon to a crosshair, or leave the blue straight arrow.

@CJTmmr : Regarding "Simulation does not show lanes or street name widget": Do not understand why this should be. Make sure that you configure these widgets to be shown for the app profile you **use for the navigation**, e.g. "car". Then, after the route has been calculated, and after tapping "Simulate", tap "Go" on the Route config dialog, and you should get everything exactly as if you were actually travelling there.

**Comment by polarbearing** [comment]
Simulation in nightly 11-Sep-2016 01:36:

Case 1, Schnellerstr, shows the turn target correctly ahead of the junction. There is is brief flicker showing the current road in the moment of the turn. However this has to be tested in driving, simulation uses fast speed. 

Case 2, Fennstr, again the main problem is fixed. In simulation, there is a brief "away from route" (grey arrow) navigation instruction. 

Case 3, Hertastr, is a different situation, here is the turn instruction missing, probably because the angle is too soft to be recognised. I have an idea how to address that and will open a separate ticket.

**Comment by sonora** [comment]
Ok, looks like the main issue is fixed then.

Yes, case 3 is totally different: I think turn instructions for that turn are suppressed because we do that for forking sub-hierarchy roads which you are **not** supposed to take, i.e. where you simply follow the course of the main road (like tracks forking from a tertiary road or motorway exits on a motorway are only announced if you have to take them). There may occasionally be situations where from the mapping this rule seems clear, but in the "real world" things are less clear.

**Comment by CJTmmr** [comment]
@sonara Testing with Route Simulation works okay now (Had to tap "Start/Go" after starting simulation.... then turnlanes en streetname widget are shown)
Issue is fixed indeed.
Thanks for the investigation and fix.

I really like the blue-straight-arrow: it clearly indicates "follow this road... till next announcement"

**Comment by sonora** [comment]
@CJTmmr: Thanks for the feedback,  and yes, in deed, I have seen more feedback from several people who like the blue arrow now that it behaves correctly. Guess I can close this issue.

**Comment by polarbearing** [comment]
> the blue-straight-arrow: it clearly indicates "follow this road... till next announcement"

And this is exactly what it NOT means and therefore leads to confusion. The icon indicates that you ARE on this road, it is not a navigation instruction what to follow!

**Comment by sonora** [comment]
True, particularly since we also use it in the off-route case now ...
Let's see how Victor feels about this. I guess I revise my opinion and also recommend the crosshair icon again.

**Comment by polarbearing** [comment]
Should we reopen this or make a new issue so the crosshair is not forgotten?

**Comment by sonora** [comment]
It should go to our UI designer, so I suggest make it a new issue (also briefly summarizing again the main points why it should be changed, it got lost in the long thread above so even I was confused in the middle   :)   )

**Comment by polarbearing** [comment]
Goto #3092.
